eventbus/kafka/sensor/kafka_sensor.go

```diff
@@ -0,0 +1,367 @@
+
+	consumer, err := sarama.NewConsumerGroupFromClient(s.groupName, client)
+	if err != nil {
+		return err
+	}
+
+	producer, err := sarama.NewAsyncProducerFromClient(client)
+	if err != nil {
+		return err
+	}
+
+	offsetManager, err := sarama.NewOffsetManagerFromClient(s.groupName, client)
+	if err != nil {
+		return err
+	}
+
+	s.client = client
+	s.consumer = consumer
+	s.kafkaHandler = &KafkaHandler{
+		Mutex:         &sync.Mutex{},
+		Logger:        s.Logger,
+		GroupName:     s.groupName,
+		Producer:      producer,
+		OffsetManager: offsetManager,
+		TriggerTopic:  s.topics.trigger,
+		Handlers: map[string]func(*sarama.ConsumerMessage) ([]*sarama.ProducerMessage, int64, func()){
+			s.topics.event:   s.Event,
+			s.topics.trigger: s.Trigger,
+			s.topics.action:  s.Action,
+		},
+	}
+
+	return nil
+}
+
+func (s *KafkaSensor) Connect(ctx context.Context, triggerName string, depExpression string, dependencies []eventbuscommon.Dependency, atLeastOnce bool) (eventbuscommon.TriggerConnection, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	// connect only if disconnected, if ever the connection is lost
+	// the connected boolean will flip and the sensor listener will
+	// attempt to reconnect by invoking this function again
+	if !s.connected {
+		go s.Listen(ctx)
+		s.connected = true
+	}
+
+	if _, ok := s.triggers[triggerName]; !ok {
+		expr, err := govaluate.NewEvaluableExpression(strings.ReplaceAll(depExpression, "-", "\\-"))
+		if err != nil {
+			return nil, err
+		}
+
+		depMap := map[string]eventbuscommon.Dependency{}
+		for _, dep := range dependencies {
+			depMap[base.EventKey(dep.EventSourceName, dep.EventName)] = dep
+		}
+
+		s.triggers[triggerName] = &KafkaTriggerConnection{
+			KafkaConnection: base.NewKafkaConnection(s.Logger),
+			sensorName:      s.sensor.Name,
+			triggerName:     triggerName,
+			depExpression:   expr,
+			dependencies:    depMap,
+			atLeastOnce:     atLeastOnce,
+			close:           s.Close,
+			isClosed:        s.IsClosed,
+		}
+	}
+
+	return s.triggers[triggerName], nil
+}
+
+func (s *KafkaSensor) Listen(ctx context.Context) {
+	defer s.Disconnect()
+
+	for {
+		if len(s.triggers) != len(s.sensor.Spec.Triggers) || !s.triggers.Ready() {
+			s.Logger.Info("Not ready to consume, waiting...")
+			time.Sleep(3 * time.Second)
+			continue
+		}
+
+		s.Logger.Infow("Consuming", zap.Strings("topics", s.topics.List()), zap.String("group", s.groupName))
+
+		if err := s.consumer.Consume(ctx, s.topics.List(), s.kafkaHandler); err != nil {
+			// fail fast if topics do not exist
+			if err == sarama.ErrUnknownTopicOrPartition {
+				s.Logger.Fatalf(
+					"Topics do not exist. Please ensure the topics '%s' have been created, or the kafka setting '%s' is set to true.",
+					s.topics.List(),
+					"auto.create.topics.enable",
+				)
+			}
+
+			s.Logger.Errorw("Failed to consume", zap.Error(err))
+			return
+		}
+
+		if err := ctx.Err(); err != nil {
+			s.Logger.Errorw("Kafka error", zap.Error(err))
+			return
+		}
+	}
```

Comment:
That's interesting that all the logic is in the Sensor and not in the KafkaTriggerConnection - did it fit better to put it here?

Reply:
(the Trigger-processing part that is...I know the Events are generic and all)

Reply:
This implementation is top down, which is in contrast to how it works for jetstream. When an event is consumed from a kafka topic, the 3 handler functions (Event, Trigger, Action) are invoked depending on which topic the event was consumed from. These functions then use the Triggers map (part of the KafkaSensor struct) to send these messages to each trigger that requires it.

To maintain seperation, I try to keep all state in the TriggerConnection struct (the events) and the Sensor only interacts with the trigger connections by invoking the [interface functions](https://github.com/argoproj/argo-events/pull/2502/files#diff-127d6164e8d0604a26ba3abaa3c694580b746ad45138ad7e601e954a6d3f4772R13-R24) (which is a superset of the TriggerConnection interface).

Reply:
Thanks. I've looked at it further since my comment. It's all very elegant.

Reply:
Thank you :)